pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/SpoolIntegrationTest.java

```diff
@@ -117,13 +121,146 @@ public void intermediateSpool()
     JsonNode stats = jsonNode.get("stageStats");
     assertNoError(jsonNode);
     DocumentContext parsed = JsonPath.parse(stats.toString());
-    List<Map<String, Object>> stage4On3 = parsed.read("$..[?(@.stage == 3)]..[?(@.stage == 4)]");
-    Assert.assertEquals(stage4On3.size(), 1, "Stage 4 should be descended from stage 3 exactly once");
 
-    List<Map<String, Object>> stage4On7 = parsed.read("$..[?(@.stage == 7)]..[?(@.stage == 4)]");
-    Assert.assertEquals(stage4On3.size(), 1, "Stage 4 should be descended from stage 7 exactly once");
+    checkSpoolTimes(parsed, 4, 3, 1);
+    checkSpoolTimes(parsed, 4, 7, 1);
+    checkSpoolSame(parsed, 4, 3, 7);
+  }
+
+  /**
+   * Test a complex with nested spools. Don't try to understand it, just check that the spools are correct.
```

Comment:
😆 
I unfortunately only noticed this comment after I tried to decipher this crazy query and gave up :P